libgadget/tests/test_gravity.c

```diff
@@ -0,0 +1,364 @@
+
+    double fac = 1 / (r2 * r);
+    if(r < h) {
+        double h_inv = 1.0 / h;
+        double h3_inv = h_inv * h_inv * h_inv;
+        double u = r * h_inv;
+        if(u < 0.5)
+            fac = 1. * h3_inv * (10.666666666667 + u * u * (32.0 * u - 38.4));
+        else
+            fac =
+                1. * h3_inv * (21.333333333333 - 48.0 * u +
+                        38.4 * u * u - 10.666666666667 * u * u * u - 0.066666666667 / (u * u * u));
+    }
+
+    for(d = 0; d < 3; d ++) {
+        accns[3*this + d] += - dist[d] * fac * All.G * P[other].Mass;
+        accns[3*other + d] += dist[d] * fac * All.G * P[this].Mass;
+    }
+}
+
+void check_accns(double * meanerr_tot, double * maxerr_tot, double *PairAccn, double meanacc)
+{
+    double meanerr=0, maxerr=-1;
+    int i;
+    /* This checks that the short-range force accuracy is being correctly estimated.*/
+    #pragma omp parallel for reduction(+: meanerr) reduction(max: maxerr)
+    for(i = 0; i < PartManager->NumPart; i++)
+    {
+        int k;
+        for(k=0; k<3; k++) {
+            double err = fabs((PairAccn[3*i+k] - (P[i].GravPM[k] + P[i].GravAccel[k]))/meanacc);
+            meanerr += err;
+            if(maxerr < err)
+                maxerr = err;
+        }
+    }
+    MPI_Allreduce(&meanerr, meanerr_tot, 1, MPI_DOUBLE, MPI_SUM, MPI_COMM_WORLD);
+    MPI_Allreduce(&maxerr, maxerr_tot, 1, MPI_DOUBLE, MPI_MAX, MPI_COMM_WORLD);
+    int64_t tot_npart;
+    sumup_large_ints(1, &PartManager->NumPart, &tot_npart);
+
+    *meanerr_tot/= (tot_npart*3.);
+}
+
+static void find_means(double * meangrav, double * suppmean, double * suppaccns)
+{
+    int i;
+    double meanacc = 0, meanforce = 0;
+    #pragma omp parallel for reduction(+: meanacc) reduction(+: meanforce)
+    for(i = 0; i < PartManager->NumPart; i++)
+    {
+        int k;
+        for(k=0; k<3; k++) {
+            if(suppaccns)
+                meanacc += fabs(suppaccns[3*i+k]);
+            meanforce += fabs(P[i].GravPM[k] + P[i].GravAccel[k]);
+        }
+    }
+    int64_t tot_npart;
+    sumup_large_ints(1, &PartManager->NumPart, &tot_npart);
+    if(suppaccns) {
+        MPI_Allreduce(&meanacc, suppmean, 1, MPI_DOUBLE, MPI_SUM, MPI_COMM_WORLD);
+        *suppmean/= (tot_npart*3.);
+    }
+    MPI_Allreduce(&meanforce, meangrav, 1, MPI_DOUBLE, MPI_SUM, MPI_COMM_WORLD);
+    *meangrav/= (tot_npart*3.);
+}
+
+/* This checks the force on each particle using a direct summation:
+ * very slow, but accurate.
+ * Periodic boundary conditions are included by mirroring the box.*/
+static int force_direct(double ErrTolForceAcc)
+{
+    double * accn = (double *) mymalloc("accelerations", 3*sizeof(double) * PartManager->NumPart);
+    memset(accn, 0, 3 * sizeof(double) * PartManager->NumPart);
+
+    int xx, yy, zz;
+    int repeat = 1;
+    /* (slowly) compute gravitational force, accounting for periodicity by just inventing extra boxes on either side.*/
+    for(xx=-repeat; xx < repeat; xx++)
```

Comment:
Isn't this just including -1 and 0 -- is it large enough? The convergence is kind of slow if you are looking at potential, I think.

Reply:
Oops, yes. I changed this a few times, it should be larger.